src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,237 @@
+    std::optional<request_ptr> last_request() const;
+    ss::chunked_fifo<request_ptr, chunk_size> _inflight_requests;
+    ss::chunked_fifo<request_ptr, chunk_size> _finished_requests;
+    friend producer_state;
+};
+
+/// Encapsulates all the state of a producer producing batches to
+/// a single raft group. At init, the producer registers itself with
+/// producer_state_manager that manages the lifecycle of all the
+/// producers on a given shard.
+class producer_state {
+public:
+    producer_state(
+      producer_state_manager& mgr,
+      model::producer_identity id,
+      raft::group_id group,
+      ss::noncopyable_function<void()> hook)
+      : _id(id)
+      , _group(group)
+      , _parent(std::ref(mgr))
+      , _last_updated_ts(ss::lowres_system_clock::now())
+      , _post_eviction_hook(std::move(hook)) {
+        register_self();
+    }
+
+    producer_state(const producer_state&) = delete;
+    producer_state& operator=(producer_state&) = delete;
+    producer_state(producer_state&&) noexcept = delete;
+    producer_state& operator=(producer_state&& other) noexcept = delete;
+    ~producer_state() noexcept { deregister_self(); }
+
+    friend std::ostream& operator<<(std::ostream& o, const producer_state&);
+
+    /// Runs the passed async function under the op_lock scope.
+    /// Additionally does the following
+    /// - de-registers self from the manager as a pre-hook
+    /// - re-registers with the manager back after the function
+    ///   completes
+    /// This helps the manager implement a lock-free eviction approach.
+    /// - A producer_state with an inflight request is not evicted, because it
+    ///   is no longer in the list of producers tracked.
+    /// - Re-registration helps the manager track LRU-ness, re-registration
+    ///   effectively puts this producer at the end of the queue when
+    ///   considering candidates for eviction.
+    template<AcceptsUnits AsyncFunc>
+    auto run_func(AsyncFunc&& func) {
+        return ss::with_gate(
+          _gate, [this, func = std::forward<AsyncFunc>(func)]() mutable {
+              return _op_lock.get_units().then(
+                [this, f = std::forward<AsyncFunc>(func)](auto units) {
+                    unlink_self();
+                    _ops_in_progress++;
+                    return ss::futurize_invoke(f, std::move(units))
+                      .then_wrapped([this](auto result) {
+                          _ops_in_progress--;
+                          link_self();
+                          return result;
+                      });
+                });
+          });
+    }
+
+    ss::future<> shutdown_input();
+    ss::future<> evict();
+    bool is_evicted() const { return _evicted; }
+
+    /* reset sequences resets the tracking state and skips the sequence
+     * checks.*/
+    result<request_ptr> try_emplace_request(
+      const model::batch_identity&,
+      model::term_id current_term,
+      bool reset_sequences = false);
```

Comment:
`reset_sequences` is a very imperative name which focuses on what it does but gives little context on when it's used; imho it's better to name it `is_tx` or something similar to hint when it should be set